### PR TITLE
Questdb benchmark support

### PR DESCRIPTION
**Pull request by patrickSpaceSurfer**
This pull request adds support for [QuestDB](https://questdb.io/), an open source time series database. QuestDB supports ingestion over InfluxDB line protocol, making `tsbs_generate_data` compatible with the `influx` format, and `tsbs_load_questdb` very similar to `tsbs_load_influx`.

For queries, QuestDB supports SQL which is executed over an `/exec` [HTTP endpoint](https://questdb.io/docs/reference/api/rest#exec---execute-queries) on port `9000` by default, and all query types in the suite are supported using the `questdb` format for both `iot` and `cpu-only` (DevOps) use cases.

### CODE REVIEW

**Review by jonatas — CHANGES_REQUESTED** [review]
Hello @patrickSpaceSurfer! thanks for adding the questDB here!

Can you please rebase the PR at the top of the master branch and run `make fmt` or `go fmt ./...` to have the code formatted?

We also have a `scripts/full_cycle_minitest/` in which you can also add one new bash script to run a full cycle of the scripts from questDB. It would be great and also guide the next users that will try to benchmark it ;)

The coverage also degraded with introducing new files and missing tests for it, can you please add the [missing tests](https://app.codecov.io/gh/timescale/tsbs/compare/157/diff)?

File: docs/questdb.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+**`--url`** (type: `string`, default: `http://localhost:9000/`)
+
+QuestDB REST end point.
+
+**`-help`**
+
+Prints available flags and their defaults:
+
+```bash
+~/tmp/go/bin/tsbs_load_questdb -help
```
Comment (jonatas) [inline docs/questdb.md:45]:
```suggestion
tsbs_load_questdb --help
```

File: docs/questdb.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+~/tmp/go/bin/tsbs_generate_data \
+--use-case="cpu-only" --seed=123 --scale=4000 \
+--timestamp-start="2016-01-01T00:00:00Z" --timestamp-end="2016-01-01T01:00:00Z" \
+--log-interval="10s" --format="influx" > /tmp/data
+```
+
+To generate a full data set for more intensive benchmarks:
+
+```bash
+~/tmp/go/bin/tsbs_generate_data \
```
Comment (jonatas) [inline docs/questdb.md:86]:
```suggestion
tsbs_generate_data \
```

File: scripts/load/load_questdb.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+EXE_FILE_NAME=${EXE_FILE_NAME:-$(which tsbs_load_questdb)}
+if [[ -z "$EXE_FILE_NAME" ]]; then
+    echo "tsbs_load_questdb not available. It is not specified explicitly and not found in \$PATH"
+    exit 1
+fi
+
+# Load parameters - common
+DATA_FILE_NAME=${DATA_FILE_NAME:-influx-data.gz}
+DATABASE_PORT=${DATABASE_PORT:-9000}
+DATABASE_HEALTH_PORT=${DATABASE_HEALTH_PORT:-9003}
```
Comment (jonatas) [inline scripts/load/load_questdb.sh:13]: I don't find a way to find the right port using docker.
Even running:

```
docker run --network=host \
      questdb/questdb
```

Can you share  a bit more how to overcome it? I still not able to load the data.

Here is a few details of my docker output:

```

2021-05-20T13:40:32.994960Z A tcp-line-server listening on 0:9009 [fd=33]
2021-05-20T13:40:32.998807Z I server-main started
2021-05-20T13:40:32.998909Z I server-main started
2021-05-20T13:40:32.998920Z I server-main web console URL(s):

	http://172.17.0.1:9000
	http://192.168.65.4:9000
	http://192.168.65.6:9000
	http://127.0.0.1:9000

```

**Review by jonatas — CHANGES_REQUESTED** [review]
I think all the steps described in sequence can be also added as a [full_cycle_minitest](https://github.com/timescale/tsbs/tree/master/scripts/full_cycle_minitest) script that we can wrap the most common usage with a small set.

Just for curiosity, I'll leave here the results running on my Macbook M1:

### Running queries

With no load, queries runs fine:

```
Run complete after 1000 queries with 4 workers (Overall query rate 1282.49 queries/sec):
```

But with loading, I just get this error:

```
 env NUM_WORKERS=1 ./scripts/run_queries/run_queries_questdb.sh                         15:14:42
Do not have any files specified - run from default queries folder as /tmp/bulk_queries/queries_questdb*
Running queries_questdb_cpu-max-all-1__1000_4000_123_2016-01-01T00:00:00Z_2016-01-02T00:00:00Z_cpu-only.dat.gz
Added index to hostname column of cpu table
panic: Get "http://localhost:9000/exec?count=false&query=%0A%09%09SELECT%0A%09%09%09hour%28timestamp%29+AS+hour%2C%0A%09%09%09max%28usage_user%29+AS+max_usage_user%2C+max%28usage_system%29+AS+max_usage_system%2C+max%28usage_idle%29+AS+max_usage_idle%2C+max%28usage_nice%29+AS+max_usage_nice%2C+max%28usage_iowait%29+AS+max_usage_iowait%2C+max%28usage_irq%29+AS+max_usage_irq%2C+max%28usage_softirq%29+AS+max_usage_softirq%2C+max%28usage_steal%29+AS+max_usage_steal%2C+max%28usage_guest%29+AS+max_usage_guest%2C+max%28usage_guest_nice%29+AS+max_usage_guest_nice%0A%09%09FROM+cpu%0A%09%09WHERE+hostname+IN+%28%27host_3011%27%29%0A%09%09++AND+timestamp+%3E%3D+%272016-01-01T13%3A04%3A15Z%27%0A%09%09++AND+timestamp+%3C+%272016-01-01T21%3A04%3A15Z%27%0A%09%09SAMPLE+BY+1h": EOF

goroutine 21 [running]:
main.(*HTTPClient).Do(0xc000992000, 0xc0000a04d0, 0xc00019a060, 0x0, 0x14a9f40, 0xc00098fc50)
	/Users/jonatasdp/code/timescale/tsbs/cmd/tsbs_run_queries_questdb/http_client.go:79 +0x12b9
main.(*processor).ProcessQuery(0xc00009a600, 0x14b29e0, 0xc0000a04d0, 0x0, 0xc000128248, 0x1, 0x1, 0x0, 0x0)
	/Users/jonatasdp/code/timescale/tsbs/cmd/tsbs_run_queries_questdb/main.go:98 +0x65
github.com/timescale/tsbs/pkg/query.(*BenchmarkRunner).processorHandler(0xc0000a0210, 0xc0000a2490, 0xc0000be000, 0x16adbe0, 0x14ad500, 0xc00009a600, 0x0)
	/Users/jonatasdp/code/timescale/tsbs/pkg/query/benchmarker.go:194 +0x126
created by github.com/timescale/tsbs/pkg/query.(*BenchmarkRunner).Run
	/Users/jonatasdp/code/timescale/tsbs/pkg/query/benchmarker.go:156 +0x206
Running queries_questdb_cpu-max-all-8__1000_4000_123_2016-01-01T00:00:00Z_2016-01-02T00:00:00Z_cpu-only.dat.gz
panic: Get "http://localhost:9000/exec?count=false&query=%0A%09%09SELECT%0A%09%09%09hour%28timestamp%29+AS+hour%2C%0A%09%09%09max%28usage_user%29+AS+max_usage_user%2C+max%28usage_system%29+AS+max_usage_system%2C+max%28usage_idle%29+AS+max_usage_idle%2C+max%28usage_nice%29+AS+max_usage_nice%2C+max%28usage_iowait%29+AS+max_usage_iowait%2C+max%28usage_irq%29+AS+max_usage_irq%2C+max%28usage_softirq%29+AS+max_usage_softirq%2C+max%28usage_steal%29+AS+max_usage_steal%2C+max%28usage_guest%29+AS+max_usage_guest%2C+max%28usage_guest_nice%29+AS+max_usage_guest_nice%0A%09%09FROM+cpu%0A%09%09WHERE+hostname+IN+%28%27host_249%27%2C+%27host_1403%27%2C+%27host_1435%27%2C+%27host_3539%27%2C+%27host_3639%27%2C+%27host_3075%27%2C+%27host_815%27%2C+%27host_2121%27%29%0A%09%09++AND+timestamp+%3E%3D+%272016-01-01T02%3A16%3A22Z%27%0A%09%09++AND+timestamp+%3C+%272016-01-01T10%3A16%3A22Z%27%0A%09%09SAMPLE+BY+1h": EOF
```

and in the server side we have the following log until the server crashes:


```
2021-05-20T18:15:32.424782Z I i.q.g.SqlCompiler plan [q=`select-group-by hour, max(usage_user) max_usage_user, max(usage_system) max_usage_system, max(usage_idle) max_usage_idle, max(usage_nice) max_usage_nice, max(usage_iowait) max_usage_iowait, max(usage_irq) max_usage_irq, max(usage_softirq) max_usage_softirq, max(usage_steal) max_usage_steal, max(usage_guest) max_usage_guest, max(usage_guest_nice) max_usage_guest_nice from (select-virtual [hour(timestamp) hour, usage_user, usage_system, usage_idle, usage_nice, usage_iowait, usage_irq, usage_softirq, usage_steal, usage_guest, usage_guest_nice] hour(timestamp) hour, usage_user, usage_system, usage_idle, usage_nice, usage_iowait, usage_irq, usage_softirq, usage_steal, usage_guest, usage_guest_nice, timestamp from (select [timestamp, usage_user, usage_system, usage_idle, usage_nice, usage_iowait, usage_irq, usage_softirq, usage_steal, usage_guest, usage_guest_nice, hostname] from cpu timestamp (timestamp) where hostname in 'host_321' and timestamp >= '2016-01-01T12:10:51Z' and timestamp < '2016-01-01T20:10:51Z')) sample by 1h`, fd=39]
2021-05-20T18:15:32.451540Z I i.q.c.h.p.JsonQueryProcessorState [39] execute-new [skip: 0, stop: 9223372036854775807]
2021-05-20T18:15:32.451766Z I i.q.c.h.p.JsonQueryProcessorState [39] timings [compiler: 27331584, count: 0, execute: 27588916, q=`
		SELECT
			hour(timestamp) AS hour,
			max(usage_user) AS max_usage_user, max(usage_system) AS max_usage_system, max(usage_idle) AS max_usage_idle, max(usage_nice) AS max_usage_nice, max(usage_iowait) AS max_usage_iowait, max(usage_irq) AS max_usage_irq, max(usage_softirq) AS max_usage_softirq, max(usage_steal) AS max_usage_steal, max(usage_guest) AS max_usage_guest, max(usage_guest_nice) AS max_usage_guest_nice
		FROM cpu
		WHERE hostname IN ('host_321')
		  AND timestamp >= '2016-01-01T12:10:51Z'
		  AND timestamp < '2016-01-01T20:10:51Z'
		SAMPLE BY 1h`]
2021-05-20T18:15:32.451797Z I i.q.c.h.p.JsonQueryProcessor all sent [fd=39, lastRequestBytesSent=1279, nCompletedRequests=1086, totalBytesSent=1046658]
2021-05-20T18:15:32.451818Z I i.q.c.h.p.QueryCache push [thread=questdb-worker-3, sql=
		SELECT
			hour(timestamp) AS hour,
			max(usage_user) AS max_usage_user, max(usage_system) AS max_usage_system, max(usage_idle) AS max_usage_idle, max(usage_nice) AS max_usage_nice, max(usage_iowait) AS max_usage_iowait, max(usage_irq) AS max_usage_irq, max(usage_softirq) AS max_usage_softirq, max(usage_steal) AS max_usage_steal, max(usage_guest) AS max_usage_guest, max(usage_guest_nice) AS max_usage_guest_nice
		FROM cpu
		WHERE hostname IN ('host_321')
		  AND timestamp >= '2016-01-01T12:10:51Z'
		  AND timestamp < '2016-01-01T20:10:51Z'
		SAMPLE BY 1h]
2021-05-20T18:15:32.453085Z I i.q.c.h.p.JsonQueryProcessorState [39] exec [q='
		SELECT
			hour(timestamp) AS hour,
			max(usage_user) AS max_usage_user, max(usage_system) AS max_usage_system, max(usage_idle) AS max_usage_idle, max(usage_nice) AS max_usage_nice, max(usage_iowait) AS max_usage_iowait, max(usage_irq) AS max_usage_irq, max(usage_softirq) AS max_usage_softirq, max(usage_steal) AS max_usage_steal, max(usage_guest) AS max_usage_guest, max(usage_guest_nice) AS max_usage_guest_nice
		FROM cpu
		WHERE hostname IN ('host_3011')
		  AND timestamp >= '2016-01-01T13:04:15Z'
		  AND timestamp < '2016-01-01T21:04:15Z'
		SAMPLE BY 1h']
2021-05-20T18:15:32.453102Z I i.q.c.h.p.QueryCache miss [thread=questdb-worker-3, sql=
		SELECT
			hour(timestamp) AS hour,
			max(usage_user) AS max_usage_user, max(usage_system) AS max_usage_system, max(usage_idle) AS max_usage_idle, max(usage_nice) AS max_usage_nice, max(usage_iowait) AS max_usage_iowait, max(usage_irq) AS max_usage_irq, max(usage_softirq) AS max_usage_softirq, max(usage_steal) AS max_usage_steal, max(usage_guest) AS max_usage_guest, max(usage_guest_nice) AS max_usage_guest_nice
		FROM cpu
		WHERE hostname IN ('host_3011')
		  AND timestamp >= '2016-01-01T13:04:15Z'
		  AND timestamp < '2016-01-01T21:04:15Z'
		SAMPLE BY 1h]
2021-05-20T18:15:32.453548Z I i.q.g.SqlCompiler plan [q=`select-group-by hour, max(usage_user) max_usage_user, max(usage_system) max_usage_system, max(usage_idle) max_usage_idle, max(usage_nice) max_usage_nice, max(usage_iowait) max_usage_iowait, max(usage_irq) max_usage_irq, max(usage_softirq) max_usage_softirq, max(usage_steal) max_usage_steal, max(usage_guest) max_usage_guest, max(usage_guest_nice) max_usage_guest_nice from (select-virtual [hour(timestamp) hour, usage_user, usage_system, usage_idle, usage_nice, usage_iowait, usage_irq, usage_softirq, usage_steal, usage_guest, usage_guest_nice] hour(timestamp) hour, usage_user, usage_system, usage_idle, usage_nice, usage_iowait, usage_irq, usage_softirq, usage_steal, usage_guest, usage_guest_nice, timestamp from (select [timestamp, usage_user, usage_system, usage_idle, usage_nice, usage_iowait, usage_irq, usage_softirq, usage_steal, usage_guest, usage_guest_nice, hostname] from cpu timestamp (timestamp) where hostname in 'host_3011' and timestamp >= '2016-01-01T13:04:15Z' and timestamp < '2016-01-01T21:04:15Z')) sample by 1h`, fd=39]
```

### Bulk loading
```
Summary:
loaded 28800000 metrics in 29.921sec with 8 workers (mean rate 962525.68 metrics/sec)
loaded 2880000 rows in 29.921sec with 8 workers (mean rate 96252.57 rows/sec)
```

I tried to adjust the parameters to properly run it on docker and I'm not sure how to debug the issue to make it work. Here are my docker settings:

Screen Shot 2021-05-20 at 15 20 47

I'll be tweaking the memory and CPU cores to see if it helps.

@patrickSpaceSurfer can you please rebase and help me to properly test it?

**Comment by jonatas** [comment]
@patrickSpaceSurfer just a heads up that increasing the size of the machine to 6 cores and 4 GB of RAM makes everything works as expected 🎉 

Here is full_mini_cycle I created to test until it works:

```bash
FORMATS=influx TS_END=2016-01-01T02:00:00Z bash ./scripts/generate_data.sh
NUM_WORKERS=4
./scripts/load/load_questdb.sh
./scripts/run_queries/run_queries_questdb.sh
```

**Comment by patrickSpaceSurfer** [comment]
> Hello @patrickSpaceSurfer! thanks for adding the questDB here!
> 
> Can you please rebase the PR at the top of the master branch and run `make fmt` or `go fmt ./...` to have the code formatted?
> 
> We also have a `scripts/full_cycle_minitest/` in which you can also add one new bash script to run a full cycle of the scripts from questDB. It would be great and also guide the next users that will try to benchmark it ;)
> 
> The coverage also degraded with introducing new files and missing tests for it, can you please add the [missing tests](https://app.codecov.io/gh/timescale/tsbs/compare/157/diff)?

I have merged the latest master, reformatted the code and added a minitest script.

With regards to coverage, I did add some new tests a few days ago, but it looks like the coverage report has not run since then. How do I run it?

**Comment by patrickSpaceSurfer** [comment]
> @patrickSpaceSurfer you can run locally [make coverage](https://github.com/timescale/tsbs/blob/master/Makefile#L44).

I have added more tests, coverage is now similar to other databases
```
pim@patrick$ > make coverage | grep quest
ok      github.com/timescale/tsbs/cmd/tsbs_generate_queries/databases/questdb   0.057s  coverage: 98.5% of statements
ok      github.com/timescale/tsbs/cmd/tsbs_load_questdb 0.139s  coverage: 54.2% of statements
?       github.com/timescale/tsbs/cmd/tsbs_run_queries_questdb  [no test files]
ok      github.com/timescale/tsbs/pkg/targets/questdb   0.033s  coverage: 75.0% of statements
```

**Review by jonatas — APPROVED** [review]
Very good @patrickSpaceSurfer!

Tested locally and all running smoothly! I hope @RobAtticus can merge it soon.

Here are the mini-cycle results running on my M1 with docker.

```
QuestDB last row per host: 1000 points
QuestDB max of all CPU metrics, random    1 hosts, random 8h0m0s by 1h: 1000 points
QuestDB CPU over threshold, 1 host(s): 1000 points
QuestDB 5 cpu metric(s), random    1 hosts, random 1h0m0s by 1m: 1000 points
QuestDB max cpu over last 5 min-intervals (random end): 1000 points
time,per. metric/s,metric total,overall metric/s,per. row/s,row total,overall row/s

Summary:
loaded 864000 metrics in 1.364sec with 1 workers (mean rate 633638.08 metrics/sec)
loaded 86400 rows in 1.364sec with 1 workers (mean rate 63363.81 rows/sec)
Run complete after 10 queries with 1 workers (Overall query rate 249.59 queries/sec):
QuestDB last row per host:
min:     1.47ms, med:     2.27ms, mean:     3.85ms, max:   17.27ms, stddev:     4.49ms, sum:   0.0sec, count: 10
all queries              :
min:     1.47ms, med:     2.27ms, mean:     3.85ms, max:   17.27ms, stddev:     4.49ms, sum:   0.0sec, count: 10
wall clock time: 0.047051sec
Run complete after 10 queries with 1 workers (Overall query rate 69.94 queries/sec):
QuestDB max of all CPU metrics, random    1 hosts, random 8h0m0s by 1h:
min:    11.80ms, med:    12.80ms, mean:    14.06ms, max:   22.71ms, stddev:     3.03ms, sum:   0.1sec, count: 10
all queries                                                           :
min:    11.80ms, med:    12.80ms, mean:    14.06ms, max:   22.71ms, stddev:     3.03ms, sum:   0.1sec, count: 10
wall clock time: 0.154872sec
Run complete after 10 queries with 1 workers (Overall query rate 236.31 queries/sec):
QuestDB CPU over threshold, 1 host(s):
min:     1.60ms, med:     2.97ms, mean:     4.03ms, max:    6.85ms, stddev:     1.92ms, sum:   0.0sec, count: 10
all queries                          :
min:     1.60ms, med:     2.97ms, mean:     4.03ms, max:    6.85ms, stddev:     1.92ms, sum:   0.0sec, count: 10
wall clock time: 0.047333sec
Run complete after 10 queries with 1 workers (Overall query rate 302.92 queries/sec):
QuestDB 5 cpu metric(s), random    1 hosts, random 1h0m0s by 1m:
min:     1.37ms, med:     2.52ms, mean:     3.23ms, max:    5.79ms, stddev:     1.35ms, sum:   0.0sec, count: 10
all queries                                                    :
min:     1.37ms, med:     2.52ms, mean:     3.23ms, max:    5.79ms, stddev:     1.35ms, sum:   0.0sec, count: 10
wall clock time: 0.038431sec
Run complete after 10 queries with 1 workers (Overall query rate 342.08 queries/sec):
QuestDB max cpu over last 5 min-intervals (random end):
min:     1.75ms, med:     2.46ms, mean:     2.64ms, max:    4.48ms, stddev:     0.84ms, sum:   0.0sec, count: 10
all queries                                           :
min:     1.75ms, med:     2.46ms, mean:     2.64ms, max:    4.48ms, stddev:     0.84ms, sum:   0.0sec, count: 10
wall clock time: 0.033503sec
```

**Comment by bsmth** [comment]
Hi @jonatas, @RobAtticus thanks for the assistance with the PR so far. Is there anything else outstanding or is this ready to be merged? Thanks a lot! 🙌🏻

**Pull request merged**